### ISSUE TITLE
[Bug] Parameter Clipping Causes Loss of AdamState in Optimizer

### ISSUE THREAD

**Issue by L-M-Sherlock** [tracel-ai/burn#2428]
### Description
When implementing parameter clipping after the optimization step, I encountered a bug that causes the loss of AdamState in the optimizer. This occurs because the clipping operation creates a new `Param`, changing the `id` of `model.w`. Since the optimizer's `AdamState` is stored in `records` with the parameter `id` as the key, this clipping step leads to the loss of the `AdamState`, preventing the optimizer from working as expected.

### Code Example
```rust
model = optim.step(lr, model, grads);
model.w = Param::from_tensor(parameter_clipper(model.w.val()));
```

### Expected Behavior
The parameter clipping should maintain the optimizer state (AdamState) for the clipped parameters.

### Actual Behavior
The optimizer state is lost after clipping, causing the optimization process to behave incorrectly in subsequent steps.

### Environment
- burn version: 0.13.1

### Possible Solution
We need a way to clip parameters without creating a new `Param` instance, or a method to update the optimizer's state when parameter IDs change.

### Related Issues
This issue is related to the feature request for a Weight Clipper: 
- #662

### Additional Context
This bug was discovered while implementing a weight clipping feature similar to PyTorch's approach, where weights are clamped after the optimization step.

### Some Output during debugging

```
// before clipping the parameters:
[src/training.rs:530:9] &model.w = Param: af0d7072-c4c1-48a7-b5e2-ece7efaf3291
// after clipping the parameters:
[src/training.rs:532:9] &model.w = Param: 20fe1431-dc2d-42a4-89f9-ffd760a9cb5f
// the optimizer records
{
  "metadata": {
    "float": "f32",
    "int": "i32",
    "format": "burn_core::record::file::PrettyJsonFileRecorder<burn_core::record::settings::FullPrecisionSettings>",
    "version": "0.13.1",
    "settings": "FullPrecisionSettings"
  },
  "item": {
    "af0d7072-c4c1-48a7-b5e2-ece7efaf3291": {
      "V1": {
        "Rank1": {
          "momentum": {
            "time": 1,
            "moment_1": {
              "value": [
                -0.005832682,
                -0.00068246666,
                -0.000255784,
                0.0010539346,
                -0.0051281727,
                0.13642916,
                0.008365896,
                -0.09502305,
                0.053447265,
                -0.28928882,
                0.05141639,
                -0.0013069924,
                0.004190517,
                -0.011830096,
                -0.00009237357,
                -0.014452745,
                0.020237459,
                0.02141046,
                0.0032307208
              ],
              "shape": [
                19
              ]
            },
            "moment_2": {
              "value": [
                0.000003401972,
                4.657545e-8,
                6.5424572e-9,
                1.1107632e-7,
                0.00000262978,
                0.0018612668,
                0.000006998727,
                0.000902926,
                0.0002856572,
                0.0083686905,
                0.00026436098,
                1.7082061e-7,
                0.0000017560199,
                0.00001399493,
                8.532763e-10,
                0.000020887906,
                0.000040954925,
                0.00004584017,
                0.0000010437418
              ],
              "shape": [
                19
              ]
            }
          }
        }
      }
    }
  },
  "_b": null
}
```

**Comment by L-M-Sherlock** [comment]
I figure it out: #1859

```rust
// read:
let (id, val) = param.comsume()
// "write":
let param = ParamID::new(id, val);
```

It fixes my issue. However, it's important to note that creating a new parameter might cause unexpected optimizer behavior.
